source/h2_stream.c

```diff
@@ -358,6 +416,35 @@ static void s_stream_destroy(struct aws_http_stream *stream_base) {
     aws_mem_release(stream->base.alloc, stream);
 }
 
+void aws_h2_stream_on_closed(struct aws_h2_stream *stream, int error_code) {
+    struct aws_linked_list pending_writes;
+    aws_linked_list_init(&pending_writes);
+    { /* BEGIN CRITICAL SECTION */
+        /* clean up any pending writes */
+        s_lock_synced_data(stream);
+        /* The stream is complete now, this will prevent further writes from being queued */
+        stream->synced_data.api_state = AWS_H2_STREAM_API_STATE_COMPLETE;
```

Comment:
In every code path we traced the state was already set
I don't like it getting set again unnecessarily

Reply:
here's an idea I like right now but feels a bit sketchy?

- DON'T grab the lock, the stream should be complete, the things we touch here should not be allowed to change anymore. leave a big comment about this though because it's weird
- ASSERT that api_state is complete or init